### ISSUE TITLE
2-way ssl support between the Org/env and the k8s-router 

### ISSUE THREAD

**Issue by madhurranjan** [30x/dispatcher#20]
We'd ideally want the request that comes in from the upstream org env to hit the k8s-router layer using https.  We'll have to have the following configuration in the nginx server block:

server {
       listen 443 ssl;
        ssl_ciphers  HIGH:!aNULL:!MD5:!DH+3DES:!kEDH;
        ssl_protocols TLSv1 TLSv1.1 TLSv1.2;
        ssl_certificate <ssl cert path>;
        ssl_certificate_key <ssl cert key>;
        ssl_client_certificate <ssl cert path>;
        ssl_verify_client on;
}

We can generate a self signed cert to test this out at the nginx layer:
Below commands would help you generate a self signed cert.

openssl req -new -newkey rsa:2048 -nodes -sha256 -keyout server.key -out server.csr
openssl x509 -in server.csr -out server.crt -req -signkey server.key -sha256 -days 365

**Comment by AdamMagaluk** [comment]
My proposal is to extend the `routing` secret that is used today.

Today the secret looks like:

```
apiVersion: v1
kind: Secret
metadata:
  name: routing
type: Opaque
data:
  api-key: abc123
```

This would enable the verification of the header `X-ROUTING-API-KEY` == `abc123`.

I propose we add `certificate.crt`, `certificate.key`, `client.crt (optional)` as data fields on the secret that would enable SSL only for all hosts in the namespace. If `client.crt` is present it would also enable client certificate verification using that cert.

Example:

```
apiVersion: v1
kind: Secret
metadata:
  name: routing
type: Opaque
data:
  api-key: abc123
  certficate.crt:  "-----BEGIN CERTIFICATE----- XYZ....
  certficate.key: "-----BEGIN PRIVATE KEY----- XYZ....
  client.crt: "-----BEGIN CERTIFICATE----- XYZ....
```
That configuration would enable both SSL/Client verification along with the header check on all hosts within the namespace.

Few questions:

- What abut specifying other SSL parameters like ciphers/protocols?
- Would enabling SSL turn off port http/80 for that host?

**Comment by mpnally** [comment]
I like this. Should we do something to make it more explicit that "cert": "turbo-apigee-ssl.crt" is referring to a key called turbo-apigee-ssl.crt in a secret called routing? For env-vars, Kubernetes would have done something like: 
{"cert": {"valueFrom": {"secretKeyRef" : {"name":"routing", "key":"turbo-apigee-ssl.crt"}}}}
It's a bit clumsy, but it is explicit, and consistent with other parts of K8S

**Comment by AdamMagaluk** [comment]
So i'm for making it more explicit however having it in secrets other than `routing` causes i think more implementation issues than it provides value.

Internally dispatcher caches k8s resources it needs so pods/ns that have the `dispatcher.routing` and secrets with the configured `routing` name. Having references to any secret name may force us to cache all secrets in the event that a namespace references it. Again it could be done but i think it's more work then it solves.

Could be something like you proposed but without the `name` eg:

```json
{"cert": {"valueFrom": {"secretKeyRef" : {"key":"turbo-apigee-ssl.crt"}}}}
```

**Comment by mpnally** [comment]
Sounds good

**Comment by AdamMagaluk** [comment]
@mpnally @whitlockjc After going through the nginx config i've realized we may have issues using multiple certificates in each of the virtual hosts that map to each hostname for a namespace. Without SNI SSL extension a single ip address can only support one ssl certificate, therefore in our configuration the first cert in the nginx config would be used for all other virtual hostnames today.

The SNI SSL extension gives support to send the hostname during the ssl handshake and allows nginx to use separate certs for each virtual host.

If we go down the current approach it would force all clients to be SNI enabled. It looks like Edge supports that but needs to be explicitly enabled [see](http://docs.apigee.com/private-cloud/content/using-sni-edge). Or should we allow a single cert/key for all of the nginx configuration at a command line level.

**Comment by mpnally** [comment]
I like SNI. It seems to me to be fundamental to the intent of HTTP 1.1,
which introduced the Host header expressly to allow multiple hostnames at
the same IP address. The fact the TLS without SNI defeated the whole intent
on the Host header seems to me like a simply a bug. I'm fine with requiring
SNI.

**Comment by AdamMagaluk** [comment]
Make sense, ill move forward with using separate certs for each virtual host.

**Comment by AdamMagaluk** [comment]
Implemented in #23
